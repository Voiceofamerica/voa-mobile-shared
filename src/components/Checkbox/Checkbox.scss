
@import '../../commonStyles/index.scss';

$checkbox-size: 30px;
$grid-size: $checkbox-size / 15;

.checkbox {
  font-size: $checkbox-size;
  cursor: pointer;

  .iconHolder {
    position: relative;
    top: $grid-size;
    margin-right: $grid-size * 4;
  }

  .content {
    flex: 1;
  }

  &.styleBox {
    display: flex;
    flex-direction: row;

    .iconHolder {
      .checkmark {
        position: absolute;
        top: $grid-size / 2;
        left: $grid-size * 2;
      }
    }
  }

  &.styleFillBox {
    display: flex;
    flex-direction: row;
  }

  &.styleItem {
    @include touchable();

    display: flex;
    flex-direction: row;
    padding: $grid-size * 2;
    border-radius: $grid-size * 4;
    overflow: hidden;
    width: auto;
    margin-top: $grid-size * 2;

    .content {
      display: flex;
      flex-direction: column;
      align-items: center;
      justify-content: center;
      user-select: none;
    }
  }
}
