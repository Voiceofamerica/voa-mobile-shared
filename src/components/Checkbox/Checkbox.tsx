
import * as React from 'react'

import SvgIcon from '../SvgIcon'
import { themed, ThemeProps, DEFAULT_THEME } from '../ThemeProvider'

import {
  checkbox,
  iconHolder,
  content,
  styleBox,
  checkmark,
  styleFillBox,
  styleItem,
} from './Checkbox.scss'

export type CheckboxType = 'box' | 'fill-box' | 'item'

const BOX: CheckboxType = 'box'
const FILL_BOX: CheckboxType = 'fill-box'
const ITEM: CheckboxType = 'item'

export interface Props extends ThemeProps {
  type?: CheckboxType
  checked?: boolean
  onChange?: (value: boolean) => void
  className?: string
  style?: React.CSSProperties
}

@themed
class Checkbox extends React.Component<Props> {
  render () {
    const { type = ITEM } = this.props

    if (type === BOX) {
      return this.renderBox()
    } else if (type === FILL_BOX) {
      return this.renderFillBox()
    } else if (type === ITEM) {
      return this.renderItem()
    } else {
      return this.renderItem()
    }
  }

  private renderBox = () => {
    const { className = '', style, children } = this.props

    return (
      <div className={`${checkbox} ${styleBox} ${className}`} onClick={this.onClick}>
        {this.renderBoxIcon()}
        <div className={content}>
          {children}
        </div>
      </div>
    )
  }

  private renderBoxIcon = () => {
    const { checked = false, theme = DEFAULT_THEME } = this.props
    const {
      checkboxCheckedColor,
      checkboxUncheckedColor,
    } = theme

    if (checked) {
      return (
        <div className={iconHolder} style={{ color: checkboxCheckedColor }}>
          <SvgIcon src='checkedOutline' />
          <SvgIcon src='check' className={checkmark} />
        </div>
      )
    } else {
      return (
        <div className={iconHolder} style={{ color: checkboxUncheckedColor }}>
          <SvgIcon src='unChecked' />
        </div>
      )
    }
  }

  private renderFillBox = () => {
    const { className = '', style, children } = this.props

    return (
      <div className={`${checkbox} ${styleFillBox} ${className}`} onClick={this.onClick}>
        {this.renderFillBoxIcon()}
        <div className={content}>
          {children}
        </div>
      </div>
    )
  }

  private renderFillBoxIcon = () => {
    const { checked = false, theme = DEFAULT_THEME } = this.props
    const {
      checkboxCheckedColor,
      checkboxUncheckedColor,
    } = theme

    if (checked) {
      return (
        <div className={iconHolder} style={{ color: checkboxCheckedColor }}>
          <SvgIcon src='checked' />
        </div>
      )
    } else {
      return (
        <div className={iconHolder} style={{ color: checkboxUncheckedColor }}>
          <SvgIcon src='unChecked' />
        </div>
      )
    }
  }

  private renderItem = () => {
    const { checked = false, className = '', style = {}, children, theme = DEFAULT_THEME } = this.props
    const {
      checkboxCheckedColor,
      checkboxItemCheckedTextColor,
      checkboxUncheckedColor,
      checkboxItemUncheckedTextColor,
    } = theme

    const color = checked ? checkboxItemCheckedTextColor : checkboxItemUncheckedTextColor
    const background = checked ? checkboxCheckedColor : checkboxUncheckedColor

    return (
      <div className={`${checkbox} ${styleItem} ${className}`} onClick={this.onClick} style={{ color, background, ...style }}>
        <div className={content}>
          {children}
        </div>
      </div>
    )
  }

  private onClick = () => {
    if (this.props.onChange) {
      this.props.onChange(!this.props.checked)
    }
  }
}

export default Checkbox
