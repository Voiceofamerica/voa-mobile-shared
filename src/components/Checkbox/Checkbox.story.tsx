
import * as React from 'react'
import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'
import { select, boolean, text, withKnobs } from '@storybook/addon-knobs'

import Checkbox, { CheckboxType } from './Checkbox'

const checkboxOptions = [
  'First',
  'Second',
  'Third',
]

class CheckboxGroup extends React.Component<{}, { checked: string[] }> {
  state = {
    checked: [checkboxOptions[1]],
  }

  render () {
    const { checked } = this.state
    const typeSelect = select<CheckboxType>('type', {
      'box': 'Default',
      'fill-box': 'Filled box',
      'item': 'Item as marker',
    }, 'box')

    return (
      <div>
        {
          checkboxOptions.map(val => (
            <Checkbox key={val} checked={checked.indexOf(val) >= 0} type={typeSelect} onChange={this.toggle(val)}>
              {val}
            </Checkbox>
          ))
        }
      </div>
    )
  }

  private toggle = (val: string) => () => {
    const { checked } = this.state

    if (checked.indexOf(val) >= 0) {
      this.setState({ checked: checked.filter(v => v !== val) })
    } else {
      this.setState({ checked: [...checked, val] })
    }
  }
}

class RadioGroup extends React.Component<{}, { checked: string }> {
  state = {
    checked: '',
  }

  render () {
    const { checked } = this.state
    const typeSelect = select<CheckboxType>('type', {
      'box': 'Default',
      'fill-box': 'Filled box',
      'item': 'Item as marker',
    }, 'box')

    return (
      <div>
        {
          checkboxOptions.map(val => (
            <Checkbox key={val} checked={checked === val} type={typeSelect} onChange={this.toggle(val)}>
              {val}
            </Checkbox>
          ))
        }
      </div>
    )
  }

  private toggle = (checked: string) => () => {
    this.setState({ checked })
  }
}

storiesOf(Checkbox.name, module)
  .addDecorator(withKnobs)
  .add('display', () => {
    const typeSelect = select<CheckboxType>('type', {
      'box': 'Default',
      'fill-box': 'Filled box',
      'item': 'Item as marker',
    }, 'box')

    return (
      <Checkbox checked={boolean('checked', false)} type={typeSelect}>{text('content', 'Check me')}</Checkbox>
    )
  })
  .add('checkbox group', () => <CheckboxGroup />)
  .add('radio group', () => <RadioGroup />)
