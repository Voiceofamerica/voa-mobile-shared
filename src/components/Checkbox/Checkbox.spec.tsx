
import * as React from 'react'
import { create } from 'react-test-renderer'
import { mount } from 'enzyme'

import Checkbox, { CheckboxType } from './Checkbox'

const allTypes: CheckboxType[] = [
  'box',
  'fill-box',
  'item',
]

jest.mock('../ThemeProvider')
jest.mock('../SvgIcon')

describe(`<${Checkbox.name} />`, () => {
  describe('shapshots', () => {
    it('should render unchecked', () => {
      const element = create((
        <Checkbox />
      ))

      expect(element).toMatchSnapshot()
    })
    it('should render checked', () => {
      const element = create((
        <Checkbox checked />
      ))

      expect(element).toMatchSnapshot()
    })
    it('should render with className', () => {
      const element = create((
        <Checkbox className='myClass' />
      ))

      expect(element).toMatchSnapshot()
    })
    it('should render with styles', () => {
      const element = create((
        <Checkbox style={{ color: 'myColor' }} />
      ))

      expect(element).toMatchSnapshot()
    })

    allTypes.forEach(type => {
      it(`should render unchecked with type ${type}`, () => {
        const element = create((
          <Checkbox type={type} />
        ))

        expect(element).toMatchSnapshot()
      })
      it(`should render checked with type ${type}`, () => {
        const element = create((
          <Checkbox type={type} checked />
        ))

        expect(element).toMatchSnapshot()
      })
      it(`should render with className and type ${type}`, () => {
        const element = create((
          <Checkbox type={type} className='myClass' />
        ))

        expect(element).toMatchSnapshot()
      })
      it(`should render with styles and type ${type}`, () => {
        const element = create((
          <Checkbox type={type} style={{ color: 'myColor' }} />
        ))

        expect(element).toMatchSnapshot()
      })
    })
  })

  describe('onChange', () => {
    it('should be called whenever the element is clicked', () => {
      const onChange = jest.fn()

      const item = mount((
        <Checkbox onChange={onChange} />
      ))

      item.simulate('click')

      expect(onChange).toHaveBeenCalledTimes(1)
    })
  })
})
