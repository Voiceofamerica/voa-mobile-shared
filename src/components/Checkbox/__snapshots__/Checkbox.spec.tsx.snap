// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> shapshots should render checked 1`] = `
<div
  className="checkbox styleBox "
  onClick={[Function]}
>
  <div
    className="iconHolder"
    style={
      Object {
        "color": undefined,
      }
    }
  >
    <svg-icon
      props={
        Object {
          "src": "checkedOutline",
        }
      }
    />
    <svg-icon
      props={
        Object {
          "className": "checkmark",
          "src": "check",
        }
      }
    />
  </div>
  <div
    className="content"
  />
</div>
`;

exports[`<Checkbox /> shapshots should render checked with type box 1`] = `
<div
  className="checkbox styleBox "
  onClick={[Function]}
>
  <div
    className="iconHolder"
    style={
      Object {
        "color": undefined,
      }
    }
  >
    <svg-icon
      props={
        Object {
          "src": "checkedOutline",
        }
      }
    />
    <svg-icon
      props={
        Object {
          "className": "checkmark",
          "src": "check",
        }
      }
    />
  </div>
  <div
    className="content"
  />
</div>
`;

exports[`<Checkbox /> shapshots should render checked with type fill-box 1`] = `
<div
  className="checkbox styleBox "
  onClick={[Function]}
>
  <div
    className="iconHolder"
    style={
      Object {
        "color": undefined,
      }
    }
  >
    <svg-icon
      props={
        Object {
          "src": "checkedOutline",
        }
      }
    />
    <svg-icon
      props={
        Object {
          "className": "checkmark",
          "src": "check",
        }
      }
    />
  </div>
  <div
    className="content"
  />
</div>
`;

exports[`<Checkbox /> shapshots should render checked with type item 1`] = `
<div
  className="checkbox styleBox "
  onClick={[Function]}
>
  <div
    className="iconHolder"
    style={
      Object {
        "color": undefined,
      }
    }
  >
    <svg-icon
      props={
        Object {
          "src": "checkedOutline",
        }
      }
    />
    <svg-icon
      props={
        Object {
          "className": "checkmark",
          "src": "check",
        }
      }
    />
  </div>
  <div
    className="content"
  />
</div>
`;

exports[`<Checkbox /> shapshots should render unchecked 1`] = `
<div
  className="checkbox styleBox "
  onClick={[Function]}
>
  <div
    className="iconHolder"
    style={
      Object {
        "color": undefined,
      }
    }
  >
    <svg-icon
      props={
        Object {
          "src": "unChecked",
        }
      }
    />
  </div>
  <div
    className="content"
  />
</div>
`;

exports[`<Checkbox /> shapshots should render unchecked with type box 1`] = `
<div
  className="checkbox styleBox "
  onClick={[Function]}
>
  <div
    className="iconHolder"
    style={
      Object {
        "color": undefined,
      }
    }
  >
    <svg-icon
      props={
        Object {
          "src": "unChecked",
        }
      }
    />
  </div>
  <div
    className="content"
  />
</div>
`;

exports[`<Checkbox /> shapshots should render unchecked with type fill-box 1`] = `
<div
  className="checkbox styleBox "
  onClick={[Function]}
>
  <div
    className="iconHolder"
    style={
      Object {
        "color": undefined,
      }
    }
  >
    <svg-icon
      props={
        Object {
          "src": "unChecked",
        }
      }
    />
  </div>
  <div
    className="content"
  />
</div>
`;

exports[`<Checkbox /> shapshots should render unchecked with type item 1`] = `
<div
  className="checkbox styleBox "
  onClick={[Function]}
>
  <div
    className="iconHolder"
    style={
      Object {
        "color": undefined,
      }
    }
  >
    <svg-icon
      props={
        Object {
          "src": "unChecked",
        }
      }
    />
  </div>
  <div
    className="content"
  />
</div>
`;

exports[`<Checkbox /> shapshots should render with className 1`] = `
<div
  className="checkbox styleBox myClass"
  onClick={[Function]}
>
  <div
    className="iconHolder"
    style={
      Object {
        "color": undefined,
      }
    }
  >
    <svg-icon
      props={
        Object {
          "src": "unChecked",
        }
      }
    />
  </div>
  <div
    className="content"
  />
</div>
`;

exports[`<Checkbox /> shapshots should render with className and type box 1`] = `
<div
  className="checkbox styleBox myClass"
  onClick={[Function]}
>
  <div
    className="iconHolder"
    style={
      Object {
        "color": undefined,
      }
    }
  >
    <svg-icon
      props={
        Object {
          "src": "unChecked",
        }
      }
    />
  </div>
  <div
    className="content"
  />
</div>
`;

exports[`<Checkbox /> shapshots should render with className and type fill-box 1`] = `
<div
  className="checkbox styleBox myClass"
  onClick={[Function]}
>
  <div
    className="iconHolder"
    style={
      Object {
        "color": undefined,
      }
    }
  >
    <svg-icon
      props={
        Object {
          "src": "unChecked",
        }
      }
    />
  </div>
  <div
    className="content"
  />
</div>
`;

exports[`<Checkbox /> shapshots should render with className and type item 1`] = `
<div
  className="checkbox styleBox myClass"
  onClick={[Function]}
>
  <div
    className="iconHolder"
    style={
      Object {
        "color": undefined,
      }
    }
  >
    <svg-icon
      props={
        Object {
          "src": "unChecked",
        }
      }
    />
  </div>
  <div
    className="content"
  />
</div>
`;

exports[`<Checkbox /> shapshots should render with styles 1`] = `
<div
  className="checkbox styleBox "
  onClick={[Function]}
>
  <div
    className="iconHolder"
    style={
      Object {
        "color": undefined,
      }
    }
  >
    <svg-icon
      props={
        Object {
          "src": "unChecked",
        }
      }
    />
  </div>
  <div
    className="content"
  />
</div>
`;

exports[`<Checkbox /> shapshots should render with styles and type box 1`] = `
<div
  className="checkbox styleBox "
  onClick={[Function]}
>
  <div
    className="iconHolder"
    style={
      Object {
        "color": undefined,
      }
    }
  >
    <svg-icon
      props={
        Object {
          "src": "unChecked",
        }
      }
    />
  </div>
  <div
    className="content"
  />
</div>
`;

exports[`<Checkbox /> shapshots should render with styles and type fill-box 1`] = `
<div
  className="checkbox styleBox "
  onClick={[Function]}
>
  <div
    className="iconHolder"
    style={
      Object {
        "color": undefined,
      }
    }
  >
    <svg-icon
      props={
        Object {
          "src": "unChecked",
        }
      }
    />
  </div>
  <div
    className="content"
  />
</div>
`;

exports[`<Checkbox /> shapshots should render with styles and type item 1`] = `
<div
  className="checkbox styleBox "
  onClick={[Function]}
>
  <div
    className="iconHolder"
    style={
      Object {
        "color": undefined,
      }
    }
  >
    <svg-icon
      props={
        Object {
          "src": "unChecked",
        }
      }
    />
  </div>
  <div
    className="content"
  />
</div>
`;
